### ISSUE TITLE
Monitoring for CodeWorld game server

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#390]
It would be good to get more visibility on what's happening inside the CodeWorld game server.

For the main web server, the current state is not so bad, for a few reasons: (a) the site is fundamentally pretty simple, (b) all useful work spawns subprocesses for GHCJS and writes artifacts with timestamps, and (c) there are logs, at least.  None of these things, though, are true of the new game server.  I'd like to know how many concurrent sessions and connections it's maintaining, how long-lived they are, whether there are leaks over time, the distribution of session sizes, the rate of messages relayed, etc.  None of these are visible in the current setup.

**Comment by nomeata** [comment]
I didn’t see this proposal yet, but some stats collectiong code is available at #403 an can of course be extended.

**Comment by cdsmith** [comment]
That looks great.  I also set up monitoring for nginx, so I now have stats on number of total TCP connections (between both servers).  Once your change is in, I'll figure out how to get Stackdriver's collectd to gather these game stats, and this will be perfect.

**Comment by nomeata** [comment]
https://collectd.org/wiki/index.php/Plugin:cURL-JSON is likely relevant.

**Comment by cdsmith** [comment]
Brainstorming some things that would be nice to see in the logging:

1. Cumulative number of games created [in #403]
2. Cumulative number of connections [in #403]
3. Cumulative number of messages [in #403]
4. Cumulative size of messages [in #403]
5. Cumulative number of dropped connections
6. Cumulative number of message parse failures.
7. Cumulative number of attempts to join non-existent games.
8. Cumulative number of attempts to join full games.
9. Number of active games in the waiting state [in #403]
10. Number of active games in the running state [in #403]
11. Number of active connections [in #403]
12. Average and maximum number of connections per game
13. Average request handling latency.

**Comment by nomeata** [comment]
5 makes no sense at this point, because every connection is dropped eventually, is it? We don't have a `quit` command or something.

**Comment by cdsmith** [comment]
Good point about dropped connections.  What we really want to know is how often the connection is dropped without the client being killed first, and server monitoring is useless for that.  Too bad.

The stats you've added are now available from the game server, by the way.  I'm still working on how to get them accessible in my monitoring consoles, as I am not an expert in collectd or stackdriver.
